kazuki-core-test/src/test/kotlin/com/anaplan/engineering/kazuki/core/test/TestSet.kt

```diff
@@ -1,87 +1,135 @@
-        assertEquals(mk_Set(mk_(3, 1), mk_(4, 1), mk_(3, 2), mk_(4, 2)), mk_Set(3, 4) x mk_Set<Int>(2, 1))
-        assertEquals(mk_Set(mk_(1, 3), mk_(1, 4), mk_(2, 3), mk_(2, 4)), mk_Set(1, 2) x mk_Set<Int>(3, 4))
+        if (allowsEmpty) {
+            assertEquals(empty(), create(1, 2, 3) x empty())
+        }
+        assertEquals(create(mk_(1, 2), mk_(1, 3)), create(1) x create(2, 3))
+        assertEquals(create(mk_(3, 1), mk_(4, 1), mk_(3, 2), mk_(4, 2)), create(3, 4) x create(2, 1))
+        assertEquals(create(mk_(1, 3), mk_(1, 4), mk_(2, 3), mk_(2, 4)), create(1, 2) x create(3, 4))
     }
 
     @Test
     fun arbitrary() {
-        assertEquals(true, mk_Set(1, 2, 3).arbitrary() in mk_Set(1, 2, 3))
-        assertEquals(false, mk_Set(1, 2, 3).arbitrary() in mk_Set(4, 5, 6))
-        assertFailsWith<PreconditionFailure> { mk_Set<Int>().arbitrary() }
+        if (allowsEmpty) {
+            causesPreconditionFailure { empty().arbitrary() }
+        }
+        assertEquals(true, create(1, 2, 3).arbitrary() in create(1, 2, 3))
+        assertEquals(false, create(1, 2, 3).arbitrary() in create(4, 5, 6))
     }
 
     @Test
     fun card() {
-        assertEquals(3, mk_Set(1, 2, 3).card)
-        assertEquals(3, mk_Set(mk_Set(1, 2), 2, 3).card)
-        assertEquals(0, mk_Set<Int>().card)
+        if (allowsEmpty) {
+            assertEquals(0, empty().card)
+        }
+        assertEquals(3, create(1, 2, 3).card)
+        assertEquals(3, create(create(1, 2), 2, 3).card)
     }
 
     @Test
     fun dunion() {
-        assertEquals(mk_Set(), dunion(mk_Set<Int>()))
-        assertEquals(mk_Set(1, 2, 3), dunion(mk_Set(mk_Set(1, 2), mk_Set(3))))
-        assertEquals(mk_Set(1, 2, 3), dunion(mk_Set(1, 2, 3)))
-        assertEquals(mk_Set(1), dunion(mk_Set(), mk_Set(1)))
-        assertEquals(mk_Set(mk_Set(1), 2, 3), dunion(mk_Set(mk_Set(1), 2), mk_Set(3)))
-        assertEquals(mk_Set(mk_Set(1, 2), 3), dunion(mk_Set(mk_Set(1, 2), 3)))
+        if (allowsEmpty) {
+            assertEquals(empty(), dunion(empty()))
+            assertEquals(create(1), dunion(empty(), create(1)))
+        }
+        assertEquals(create(1, 2, 3), dunion(mk_Set1(create(1, 2), create(3))))
+        assertEquals(create(1, 2, 3), dunion(create(1, 2, 3)))
+        assertEquals(create(create(1), 2, 3), dunion(mk_Set1(create(1), 2), create(3)))
+        assertEquals(create(create(1, 2), 3), dunion(create(create(1, 2), 3)))
     }
 
     @Test
     fun plus() {
-        assertEquals(mk_Set(1), mk_Set<Int>().plus(1))
-        assertEquals(mk_Set(1), mk_Set(1).plus(mk_Set()))
-        assertEquals(mk_Set(1, 2), mk_Set(1).plus(mk_Set(1, 2)))
-        assertEquals(mk_Set(1, 2), mk_Set(1).plus(2))
+        if (allowsEmpty) {
+            assertEquals(create(1), empty().plus(1))
+            assertEquals(create(1), create(1).plus(empty()))
+        }
+        assertEquals(create(1, 2), create(1).plus(create(1, 2)))
+        assertEquals(create(1, 2), create(1).plus(2))
```

Comment:
:pick: these should really use the operator.. it is not generally done to invoke the function directly e.g. 

` empty().plus(1)` should be `empty() + 1`